rules/store.go

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE
+
+package rules
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3metrics/generated/proto/schema"
+)
+
+var (
+	errNilRuleSet    = errors.New("nil ruleset")
+	errNilNamespaces = errors.New("nil namespaces")
+)
+
+// Store facilitates read/write operations to the backing kv store.
+type Store interface {
+	// Persist saves the given ruleset and optional namespace to the backing store.
+	// If saveNamespace is false, will only use the namesapces state to ensure that
+	// the RuleSet is still valid.
+	Persist(RuleSet, *Namespaces, bool) error
```

Comment:
RuleSet is an interface, Namespaces is not. As for why not just use the value, Handler.Namespace(...) returns a *Namespace because I think it makes more sense to return nil than a blank Namespace in the case its not found.